fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -0,0 +1,268 @@
+        enforce(event.getFhirResource(), getPatientIdFromToken(jwt), Permission.READ, getScopesFromToken(jwt));
+    }
+
+    @Override
+    public void afterVread(FHIRPersistenceEvent event) throws FHIRPersistenceInterceptorException {
+        DecodedJWT jwt = JWT.decode(getAccessToken());
+        enforce(event.getFhirResource(), getPatientIdFromToken(jwt), Permission.READ, getScopesFromToken(jwt));
+    }
+
+    @Override
+    public void afterHistory(FHIRPersistenceEvent event) throws FHIRPersistenceInterceptorException {
+        DecodedJWT jwt = JWT.decode(getAccessToken());
+        if (event.getFhirResource() instanceof Bundle) {
+            for ( Bundle.Entry entry : ((Bundle) event.getFhirResource()).getEntry()) {
+                if (entry.getResource() != null) {
+                    enforce(entry.getResource(), getPatientIdFromToken(jwt), Permission.READ, getScopesFromToken(jwt));
+                }
+            }
+        } else {
+            throw new IllegalStateException("Expected event resource of type Bundle but found " +
+                    event.getFhirResource().getClass().getSimpleName());
+        }
+    }
+
+    @Override
+    public void afterSearch(FHIRPersistenceEvent event) throws FHIRPersistenceInterceptorException {
+        DecodedJWT jwt = JWT.decode(getAccessToken());
+        if (event.getFhirResource() instanceof Bundle) {
+            for ( Bundle.Entry entry : ((Bundle) event.getFhirResource()).getEntry()) {
+                if (entry.getResource() != null) {
+                    enforce(entry.getResource(), getPatientIdFromToken(jwt), Permission.READ, getScopesFromToken(jwt));
+                }
+            }
+        } else {
+            throw new IllegalStateException("Expected event resource of type Bundle but found " +
+                    event.getFhirResource().getClass().getSimpleName());
+        }
+    }
+
+    /**
+     * @param event
+     * @param contextIds
+     * @param collect
+     * @throws FHIRPersistenceInterceptorException
+     */
+    private void enforce(Resource resource, List<String> contextIds, Permission requiredPermission, List<Scope> approvedScopes)
+            throws FHIRPersistenceInterceptorException {
+        Objects.requireNonNull(resource, "resource");
+        Objects.requireNonNull(contextIds, "contextIds");
+
+        String resourceType = resource.getClass().getSimpleName();
+        Map<ContextType, List<Scope>> approvedScopeMap = approvedScopes.stream()
+                // First filter the list to only scopes which grant the required permissions on the passed resourceType
+                .filter(s -> s.getResourceType() == ResourceType.ValueSet.RESOURCE ||
+                        s.getResourceType().value().equals(resourceType))
+                .filter(s -> hasPermission(s.getPermission(),requiredPermission))
+                // Then group the scopes by their context type
+                .collect(Collectors.groupingBy(s -> s.getContextType()));
```

Comment:
Given John's issues with streams, how does this perform?

Reply:
there will typically only be one or two scope strings in the JWT, so I'm really not worried about it